Wikipedia/Code/SavedArticlesFetcher.m

```diff
@@ -81,6 +87,29 @@ - (instancetype)initWithDataStore:(MWKDataStore *)dataStore
                   imageInfoFetcher:[[MWKImageInfoFetcher alloc] init]];
 }
 
+#pragma mark - Progress
+
+// Reminder: due to the internal structure of this class and how it is presently being used, we can't simply check the 'count' of 'fetchOperationsByArticleTitle' dictionary for the total. (It doesn't reflect the actual total.) Could re-plumb this class later.
+- (NSUInteger)calculateTotalArticlesToFetchCount {
```

Comment:
Any reason to always require the DB rather than just keep the fetch in memory and observe the changes (This may require an internal NSFRC but maybe not). 

I guess this is probably less error prone. Im not sure if the performance is a concern. 

Well actually, performance may be a concern since you are using the "viewContext". Actually, why not use a background context and then at least the UI won't ever get held up while performing the fetch.

Also is do you need to ensure you are on the main queue? Or should you use the a block from the moc to make sure you are on the right queue?

Reply:
I'll check...

Reply:
> Any reason to always require the DB rather than just keep the fetch in memory and observe the changes (This may require an internal NSFRC but maybe not).

Ya the reason is the way the `_update` method currently works, which is a bit odd and hard to follow. It's using the same context and predicate and an `updateAgain` closure to recursively continue kicking off `fetchArticleURL` calls, one by one, until they've all wrapped up. These `fetchArticleURL` calls are what actually populate a `fetchOperationsByArticleTitle` dict, but the `fetchOperationsByArticleTitle` `count` doesn't reflect the total we need - it's usually `0` or `1` because of the aforementioned structure, so we can't monitor it for the total.

> Well actually, performance may be a concern since you are using the "viewContext". Actually, why not use a background context and then at least the UI won't ever get held up while performing the fetch.

I think because I'm not fetching (inflating) the actual `WMFArticle`'s, but rather only the count matching the predicate, it's pretty snappy. On my machine it looks to be around 0.0005 seconds with only a few articles, but also around 0.0005 seconds with a hundred or so articles, so seems safe.

Re the view context, it's the same one used everywhere else within the `SaveArticlesFetcher`, (such as in its `update` method). Was hesitant to re-plumb more than necessary at this time. I can add a main thread assert (as seen in the `update` method) to make sure we're on the expected one.

Reply:
> These fetchArticleURL calls are what actually populate a fetchOperationsByArticleTitle dict, but the fetchOperationsByArticleTitle count doesn't reflect the total we need - it's usually 0 or 1 because of the aforementioned structure, so we can't monitor it for the total.

Oh I wasn't proposing monitoring the dictionary. I am asking rather than executing a DB fetch every time to update the count, why not fetch once and then update the results by monitoring changes to the MOC… BUT I think that complexity is probably not warranted given your performance numbers - so disregard :)

> I think because I'm not fetching (inflating) the actual WMFArticle's, but rather only the count matching the predicate, it's pretty snappy. On my machine it looks to be around 0.0005 seconds with only a few articles, but also around 0.0005 seconds with a hundred or so articles, so seems safe.

> Re the view context, it's the same one used everywhere else within the SaveArticlesFetcher, (such as in its update method). Was hesitant to re-plumb more than necessary at this time. 

Hmm… interesting… do we know why we are using the view context in a fetcher that I am guessing is running a "background process"? Maybe there is something I am missing here… I wouldn't want you to change this without understanding why.

Usually, if DB fetches can be done off the main thread, it is good to do so (In general, I wouldn't use the view context for fetching object unless I then had to the manipulate the fetched objects on the main thread).

Who knows what types of environmental conditions may arise that cause the fetch to be slow (old device, large DB, constrained memory, etc). So even if performance is good in your testing, is such issues come up for users in the real world, the UI may lock up and could lead to hard to debug performance issues down the road.

> I can add a main thread assert (as seen in the update method) to make sure we're on the expected one.

Depends… is this method designed only to be called from the main thread? If so, then yes. But if it is expected to be able to be called from any thread, I would use the MOC block method to ensure the work is done on the right queue.

Reply:
> Hmm… interesting… do we know why we are using the view context in a fetcher that I am guessing is running a "background process"? Maybe there is something I am missing here… I wouldn't want you to change this without understanding why.

No, I wasn't sure why it was this way. I'll take a look at the blame log for hints...